src/rnnlm/rnnlm-lattice-rescoring.cc

```diff
@@ -0,0 +1,204 @@
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include <utility>
+
+#include "rnnlm/rnnlm-lattice-rescoring.h"
+#include "util/stl-utils.h"
+#include "util/text-utils.h"
+
+namespace kaldi {
+namespace nnet3 {
+
+KaldiRnnlmDeterministicFst::~KaldiRnnlmDeterministicFst() {
+  int size = state_to_decodable_rnnlm_.size();
+  KALDI_ASSERT(state_to_nnet3_output_.size() == size);
+  for (int i = 0; i < size; i++) {
+    delete state_to_decodable_rnnlm_[i];
+    delete state_to_nnet3_output_[i];
+  }
+  
+  state_to_decodable_rnnlm_.resize(0);
+  state_to_nnet3_output_.resize(0);
+  state_to_wseq_.resize(0);
+  wseq_to_state_.clear();
+}
+
+void KaldiRnnlmDeterministicFst::Clear() {
+  int size = state_to_decodable_rnnlm_.size();
+  KALDI_ASSERT(state_to_nnet3_output_.size() == size);
+  for (int i = 1; i < size; i++) {
+    delete state_to_decodable_rnnlm_[i];
+    delete state_to_nnet3_output_[i];
+  }
+  
+  state_to_decodable_rnnlm_.resize(1);
+  state_to_nnet3_output_.resize(1);
+  state_to_wseq_.resize(1);
+  wseq_to_state_.clear();
+  wseq_to_state_[state_to_wseq_[0]] = 0;
+}
+
+void KaldiRnnlmDeterministicFst::ReadFstWordSymbolTableAndRnnWordlist(
+    const std::string &rnn_wordlist,
+    const std::string &word_symbol_table_rxfilename) {
+  // Reads symbol table.
+  fst::SymbolTable *fst_word_symbols = NULL;
+  if (!(fst_word_symbols =
+      fst::SymbolTable::ReadText(word_symbol_table_rxfilename))) {
+    KALDI_ERR << "Could not read symbol table from file "
+              << word_symbol_table_rxfilename;
+  }
+
+  full_voc_size_ = fst_word_symbols->NumSymbols();
+  fst_label_to_word_.resize(full_voc_size_);
+
+  for (int32 i = 0; i < fst_label_to_word_.size(); ++i) {
+    fst_label_to_word_[i] = fst_word_symbols->Find(i);
+    if (fst_label_to_word_[i] == "") {
+      KALDI_ERR << "Could not find word for integer " << i << "in the word "
+                << "symbol table, mismatched symbol table or you have discoutinuous "
+                << "integers in your symbol table?";
+    }
+  }
+
+  fst_label_to_rnn_label_.resize(fst_word_symbols->NumSymbols(), -1);
+
+  out_OOS_index_ = 1;
+  {
+    std::ifstream ifile(rnn_wordlist.c_str());
+    int32 id;
+    string word;
+    int32 i = 0;
+    while (ifile >> word >> id) {
+      if (word == "</s>") {
+        eos_index_ = id;
+      } else if (word == "<s>") {
+        bos_index_ = id;
+      }
+      KALDI_ASSERT(i == id);
+      i++;
+      rnn_label_to_word_.push_back(word);
+
+      int fst_label = fst_word_symbols->Find(rnn_label_to_word_[id]);
+      if (fst::SymbolTable::kNoSymbol != fst_label || id == out_OOS_index_ || id == 0 || word == "<brk>") {}
+      else {
+        KALDI_LOG << "warning: word " << word << " in RNNLM wordlist but not in FST wordlist";
+      }
+      if (id != out_OOS_index_ && out_OOS_index_ != 0 && fst_label != -1) {
```

Comment:
Capitals should not be in variables like this.
And I think it's strange that you are implementing this mapping of symbols; I think it would be best to require that they are the same.  It's easy to generate an embedding matrix w.r.t. an arbitrary wordlist if you are using the letter features; and if not, it wouldn't be that hard to write a script to mess with the row-indexes of the word-embedding matrix.  Dealing with the strings like this in the code is rather messy.